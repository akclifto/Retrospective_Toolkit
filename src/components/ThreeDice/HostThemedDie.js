import React, { useEffect } from "react";
import { useTexture } from "@react-three/drei";
import { useBox } from "@react-three/cannon";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import { useGesture, useDrag } from "react-use-gesture";

/* istanbul ignore next */
const singleRollSound = () => {
  new Audio("/diceRoll.m4a").play();
};

/* istanbul ignore next */
const ThemedDie = (props) => {
  const {
    diceInitPos,
    rerollValue,
    rerollDieToggle,
    imageSet,
    geom,
    setOrbitControl,
    mousePos,
    rotationValues,
  } = props;
  const textures = useTexture([...imageSet]);

  const [mesh, api] = useBox(() => ({
    mass: 300,
    inertia: 13,
    position: [diceInitPos[0], diceInitPos[1], diceInitPos[2]],
    rotation: [
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI,
    ],
    linearDamping: 0.5,
    angularDamping: 0.1,
    material: { restitution: 0.3 },
  }));

  const isDragging = React.useRef(false);

  const clickBind = useGesture({
    onClick: () => {
      if (!isDragging.current) {
        rerollDieToggle(!rerollValue);
        singleRollSound();
      }
    },
  });

  const dragBind = useDrag(
    ({ first, last }) => {
      if (first) {
        setOrbitControl(false);
        isDragging.current = true;
        api.mass.set(0);
        api.collisionResponse.set(0);
      } else if (last) {
        // eslint-disable-next-line no-return-assign
        setTimeout(() => (isDragging.current = false), 100);
        setOrbitControl(true);
        api.mass.set(300);
        api.collisionResponse.set(1);
      }
      api.position.set(mousePos[0], 2, -mousePos[1]);
    },
    {
      initial: () => [mousePos[0], mousePos[1]],
      filterTaps: true,
    }
  );

  useEffect(() => {
    api.position.set(diceInitPos[0], diceInitPos[1], diceInitPos[2]);
    api.velocity.set(15, 0, -10);
    api.angularVelocity.set(-15, 2, -10);
    api.rotation.set(
      rotationValues[0] * Math.PI,
      rotationValues[1] * Math.PI,
      rotationValues[2] * Math.PI
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rotationValues]);

  return (
    <mesh
      ref={mesh}
      geometry={geom}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...dragBind()}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...clickBind()}
    >
      <boxBufferGeometry />
      {textures.map((image) => (
        <meshStandardMaterial
          key={uuidv4()}
          flatShading
          roughness={0.8}
          attachArray="material"
          map={image}
        />
      ))}
    </mesh>
  );
};

ThemedDie.propTypes = {
  diceInitPos: PropTypes.arrayOf(PropTypes.number).isRequired,
  imageSet: PropTypes.arrayOf(PropTypes.string).isRequired,
  rerollValue: PropTypes.bool.isRequired,
  rerollDieToggle: PropTypes.func.isRequired,
  geom: PropTypes.shape().isRequired,
  setOrbitControl: PropTypes.func.isRequired,
  mousePos: PropTypes.arrayOf(PropTypes.number).isRequired,
  rotationValues: PropTypes.arrayOf(PropTypes.number).isRequired,
};

export default ThemedDie;
