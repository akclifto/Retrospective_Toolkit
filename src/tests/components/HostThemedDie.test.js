import React from "react";
import { shallow } from "enzyme";
import { Provider } from "jotai";

import HostThemedDie from "../../components/ThreeDice/HostThemedDie";

describe("ThreeDice/GameManager Testing", () => {
  it("should do a snapshot test on GameManager", async () => {
    const wrapper = shallow(
      <Provider>
        <HostThemedDie />
      </Provider>
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });
});
