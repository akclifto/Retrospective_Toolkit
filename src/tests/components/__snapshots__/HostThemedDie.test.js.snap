// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeDice/GameManager Testing should do a snapshot test on GameManager 1`] = `
<ContextProvider
  value={
    Object {
      Symbol(): Object {
        "l": Set {},
        "n": Object {
          "current": 0,
        },
        "u": [Function],
        "v": Object {
          "current": Object {
            "add": [Function],
            "del": [Function],
            "read": [Function],
            "write": [Function],
          },
        },
      },
    }
  }
>
  <Component
    value={
      Object {
        "a": WeakMap {},
        "m": Map {},
        "w": Map {},
      }
    }
  >
    <InnerProvider
      c={
        Object {
          "$$typeof": Symbol(react.context),
          "Provider": [Function],
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            Symbol(): Object {
              "l": Set {},
              "n": Object {
                "current": -1,
              },
              "u": [Function],
              "v": Object {
                "current": null,
              },
            },
          },
          "_currentValue2": Object {
            Symbol(): Object {
              "l": Set {},
              "n": Object {
                "current": -1,
              },
              "u": [Function],
              "v": Object {
                "current": null,
              },
            },
          },
          "_threadCount": 0,
          Symbol(): Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
        }
      }
      r={
        Object {
          "current": [Function],
        }
      }
    >
      <ThemedDie />
    </InnerProvider>
  </Component>
</ContextProvider>
`;
